server/src/main/java/org/elasticsearch/index/engine/EngineResetLock.java

```diff
@@ -0,0 +1,103 @@
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.engine;
+
+import org.elasticsearch.core.Assertions;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Reentrant read/write lock used to control accesses to a shard's engine that can be reset.
+ *
+ * Implemented as a simple wrapper around a {@link ReentrantReadWriteLock} to make it easier to add/override methods.
+ */
+public final class EngineResetLock implements ReadWriteLock {
+
+    private final ReentrantReadWriteLock lock;
+
+    public EngineResetLock() {
+        this.lock = Assertions.ENABLED ? new QueuedWriterThreadsReentrantReadWriteLock() : new ReentrantReadWriteLock();
+    }
+
+    @Override
+    public Lock writeLock() {
+        return lock.writeLock();
+    }
+
+    @Override
+    public Lock readLock() {
+        return lock.readLock();
+    }
+
+    /**
+     * See {@link ReentrantReadWriteLock#isWriteLocked()}
+     */
+    public boolean isWriteLocked() {
+        return lock.isWriteLocked();
+    }
+
+    /**
+     * See {@link ReentrantReadWriteLock#isWriteLockedByCurrentThread()}
+     */
+    public boolean isWriteLockedByCurrentThread() {
+        return lock.isWriteLockedByCurrentThread();
+    }
+
+    /**
+     * Returns {@code true} if the number of read locks held by any thread is greater than zero.
+     * This method is designed for use in monitoring system state, not for synchronization control.
+     *
+     * @return {@code true} if any thread holds a read lock and {@code false} otherwise
+     */
+    public boolean isReadLocked() {
+        return lock.getReadLockCount() > 0;
+    }
+
+    /**
+     * Returns {@code true} if the number of holds on the read lock by the current thread is greater than zero.
+     * This method is designed for use in monitoring system state, not for synchronization control.
+     *
+     * @return {@code true} if the number of holds on the read lock by the current thread is greater than zero, {@code false} otherwise
+     */
+    public boolean isReadLockedByCurrentThread() {
+        return lock.getReadHoldCount() > 0;
+    }
+
+    /**
+     * See {@link ReentrantReadWriteLock#getReadLockCount()}
+     */
+    public int getReadLockCount() {
+        return lock.getReadLockCount();
+    }
+
+    /**
+     * See {@link ReentrantReadWriteLock#getQueuedWriterThreads()}
+     */
+    public Collection<Thread> getQueuedWriterThreads() {
```

Comment:
I wonder about the package belonging of this class, seems more `IndexShard` related than `Engine`, since it protects something in `IndexShard` primarily. If we move this to `shard` then this can be package private too - I think it is only used in tests.

Reply:
Makes sense, I pushed 4fa503c01d191cbdb41ad123e5f3e3973c4e48dc